Makefile

```
@@ -1,41 +1,185 @@
+		) \
+	)
+	@[ -n "$(VARIABLE)" ] && echo -e "\n--- VARIABLES ---"  || true;
+	@$(foreach \
+		variable, \
+		$(VARIABLE), \
+		echo "$(variable)"; \
+		[ -n "$(HELP_$(variable))" ] && echo -e "\t$(HELP_$(variable))" || true; \
+		echo -e "\t(currently: $($(variable)))"; \
+	)
+
+
+TARGET += default
+HELP_default  = builds project
+.PHONY: default
+default: build
 
+TARGET += all
+HELP_all = builds, flashes and starts monitor
 .PHONY: all
-all:
-	@true
+all: build flash monitor
 
-.PHONY: setup
-setup:
-	@true
+TARGET += check
+HELP_check = checks if everything is correctly setup
+.PHONY: check
+check: check-build check-flash check-monitor
 
+TARGET += clean
+HELP_clean = cleans up project
 .PHONY: clean
-clean:
-	@true
+clean: clean-build clean-flash clean-monitor clean-dev
+	@make --no-print-directory -C docker clean
 
+TARGET += distclean
+HELP_distclean = removes all generated files
 .PHONY: distclean
-distclean:
-	@true
+distclean: clean distclean-build distclean-flash distclean-monitor distclean-dev
+	@make --no-print-directory -C docker distclean
+
+### build targets ###
 
+TARGET += build
+HELP_build = builds project
 .PHONY: build
-build:
+build: | check-build
+	@make --no-print-directory -C docker pio EXEC="pio run"
+
+.PHONY: check-build
+check-build: check-docker
+
+.PHONY: clean-build
+clean-build:
+	@make --no-print-directory -C docker pio EXEC="pio run -t clean"
+
+.PHONY: distclean-build
+distclean-build: clean-build
+	rm -rf .pio
+
+### flash targets ###
+
+TARGET += flash
+HELP_flash = flashes project to esp
+.PHONY: flash
+flash: | check-flash
+	@make --no-print-directory -C docker \
+		pio \
+		EXEC="sudo chgrp developer $(DEV); pio run --upload-port '$(DEV)' -t upload" DEV="$(DEV)"
+
+.PHONY: check-flash
+check-flash: | check-docker check-dev
+
+.PHONY: clean-flash
+clean-flash:
+	@true
+
+.PHONY: distclean-flash
+distclean-flash: clean-flash
 	@true
 
-.PHONY: tests
-tests:
+### monitor targets ###
+
+TARGET += monitor
+HELP_monitor = connects to esp32 via serial
+.PHONY: monitor
+monitor: | check-monitor
+	@make --no-print-directory -C docker \
+		pio \
+		EXEC="sudo chgrp developer $(DEV); \
+			python3 \
+				/home/developer/.platformio/packages/framework-espidf/tools/idf_monitor.py \
+				/mnt/.pio/build/pico32/firmware.elf \
+				--port $(DEV)"
+
+TARGET_monitor += check-monitor
+HELP_check-monitor = checks if monitor is usable
+.PHONY: check-monitor
+check-monitor: | check-docker check-dev
+
+.PHONY: clean-monitor
+clean-monitor:
 	@true
+
+.PHONY: distclean-monitor
+distclean-monitor: clean-monitor
+	@true
+
+### dev targets ###
+
+TARGET += dev
+HELP_dev = specifies which USB device to connect to
+.PHONY: dev
+dev:
+	echo -n "DEV=" > ./.dev; \
+	(for dev in $$(ls /dev/serial/by-id); do echo "$$(readlink -f /dev/serial/by-id/$$dev) $$dev "; done ) \
+	| dialog \
+		--stdout \
+		--menu "choose which device to flash" 0 0 0 \
+		--file /dev/stdin \
+	>> ./.dev
```

Add a clear to clear screen after make dev

```suggestion
	>> ./.dev
	clear
```